### PR TITLE
fix: recapture CUDA graphs for RL online weight updates

### PR DESCRIPTION
**Pull request by Gao016**
## Motivation

  We discovered this bug on 8×H200 GPUs while using slime (https://github.com/THUDM/slime, an RL training framework built on top of SGLang) in colocate mode to train Qwen3.5-35B-A3B — a hybrid architecture model with Gated DeltaNet (linear attention) + Gated Attention + MoE. Its hidden layout is 10 × (3 × (Gated DeltaNet → MoE) → 1 × (Gated Attention → MoE)).

  **Symptom:** Step 0 rollout is normal. Step 1 rollout (after one round of training and weight sync) produces completely garbled text. No crash or error is raised.

  **Root cause:** During CUDA graph capture, the GPU data_ptr of every tensor accessed in the forward pass is baked into the recorded graph. Model weights are updated via param.data.copy_(), which preserves the original data_ptr — so weight addresses remain valid after updates, and CUDA graphs can correctly access the new weight values.

  However, for hybrid models like Qwen3.5-35B-A3B, the linear attention layers maintain recurrent state buffers (conv_state / ssm_state) — analogous to KV cache in full attention, but accessed by direct data_ptr baked into CUDA graphs. These buffers are allocated under the GPU_MEMORY_TYPE_KV_CACHE tag in torch_memory_saver. When pause(KV_CACHE) is called, these buffers are offloaded to CPU and their GPU memory is freed. When resume(KV_CACHE) restores them, the CUDA caching allocator may assign different GPU addresses — but the CUDA graphs still replay with the old data_ptr, causing the model to read garbage data and write to invalid memory for its recurrent state. Even though the weights are correct, the model has completely lost its token-by-token "memory", producing garbled output.

  In slime's RL training loop:

  1. pause(KV_CACHE) offloads linear attention states to CPU, freeing GPU memory
  2. Training runs for minutes, heavily fragmenting GPU memory
  3. resume(KV_CACHE) restores linear attention states, but they may land at different GPU addresses
  4. CUDA graphs still replay with the old data_ptr values → garbled output

  Step 0 works because the first pause/resume cycle has minimal GPU memory activity between them, so tensors happen to restore to the same addresses. After training, GPU memory is completely reshuffled.

  As a comparison, we also tested Qwen3-30B-A3B (GQA + MoE, no linear attention) and found it works correctly, since it has no linear attention recurrent state buffers whose data_ptr would be baked into CUDA graphs.

  ## Modifications

  Mark CUDA graphs as stale after every successful weight update, then recapture in resume_memory_occupation before returning. Only weight_updater.py is modified.

  - Add mark_cuda_graphs_stale() / recapture_cuda_graphs_after_weight_update().
  - Call mark_cuda_graphs_stale() in all 4 weight update paths: update_weights_from_disk, update_weights_from_distributed, update_weights_from_tensor, update_weights_from_ipc.
  - Trigger recapture in resume_memory_occupation if the flag is set.

  ## Accuracy Tests

  Verified on Qwen3.5-35B-A3B (--cuda-graph-bs 1 2 4 8 16 ... 256) using slime's RL training loop:

  - Without fix: step 0 normal, step 1 garbled output
  - With fix: correct output across all training steps

  ## Speed Tests and Profiling

  ---

  ## Checklist

  ---

---
### CI States

Latest PR Test (Base): :x: [Run #28355926431](https://github.com/sgl-project/sglang/actions/runs/28355926431)
Latest PR Test (Extra): :x: [Run #28355926315](https://github.com/sgl-project/sglang/actions/runs/28355926315)

### CODE REVIEW

File: python/sglang/srt/managers/scheduler_components/weight_updater.py [hunk 1]
```diff
@@ -85,6 +85,15 @@ def flush_cache_after_weight_update(self, recv_req) -> None:
...
             assert flush_cache_success, "Cache flush failed after updating weights"
 
+    def recapture_cuda_graphs_after_weight_update(self) -> None:
+        model_runner = self.tp_worker.model_runner
+        if model_runner.graph_runner is not None:
+            logger.info("Recapturing CUDA graphs after weight update")
+            model_runner.graph_runner.capture()
+
+    def mark_cuda_graphs_stale(self) -> None:
+        self.cuda_graphs_need_recapture = True
```
[the remark this answers is not in the document]
Comment (Gao016) [reply python/sglang/srt/managers/scheduler_components/weight_updater.py:148]: already fixed in the latest commit

**Comment by Gao016** [comment]
hi，@Ying1123，sorry to bother you, could you please help me to review this PR？thanks.

**Comment by kaixih** [comment]
i think we saw a similar issue on GB200 RL with miles. 

cc. @maocheng23
